File: swift/70000TonsBands/AIScheduleStorage.swift
//
//  AIScheduleStorage.swift
//  70000TonsBands
//
//  Persists "has run AI schedule" per year and backup of attendance for restore.
//

import Foundation

/// Storage for AI schedule: "has run" flag per year and attendance backup for restore.
enum AIScheduleStorage {
    private static let hasRunPrefix = "AIScheduleHasRun_"
    private static let backupFileNamePrefix = "AIScheduleBackup_"
    
    static func hasRunAI(for year: Int) -> Bool {
        UserDefaults.standard.bool(forKey: hasRunPrefix + String(year))
    }
    
    static func setHasRunAI(for year: Int, value: Bool) {
        UserDefaults.standard.set(value, forKey: hasRunPrefix + String(year))
    }
    
    private static func backupFileURL(for year: Int) -> URL {
        FilePaths.directoryPath.appendingPathComponent(backupFileNamePrefix + String(year) + ".json")
    }
    
    /// Save attendance state for the given year (only keys ending with ":year").
    /// Call before applying AI schedule so we can restore later.
    static func saveBackup(attended: [String: String], year: Int) {
        let suffix = ":" + String(year)
        let filtered = attended.filter { $0.key.hasSuffix(suffix) }
        guard !filtered.isEmpty else { return }
        do {
            let data = try JSONSerialization.data(withJSONObject: filtered)
            try data.write(to: backupFileURL(for: year))
        } catch {
            print("AIScheduleStorage: Failed to save backup for year \(year): \(error)")
        }
    }

    /// Saves year slice for wizard rollback (may be empty `{}`). Always writes a file so cancel-after-clear can restore.
    static func saveWizardRollbackBackup(attended: [String: String], year: Int) {
        let suffix = ":" + String(year)
        let filtered = attended.filter { $0.key.hasSuffix(suffix) }
        do {
            let data = try JSONSerialization.data(withJSONObject: filtered)
            try data.write(to: backupFileURL(for: year))
        } catch {
            print("AIScheduleStorage: Failed to save wizard rollback backup for year \(year): \(error)")
        }
    }

    /// Restores attendance from the wizard pre-build snapshot and removes the backup file. Does not change hasRunAI.
    static func restoreWizardCancelled(attendedHandle: ShowsAttended, year: Int) {
        guard let backup = loadBackup(year: year) else { return }
        let yearStr = String(year)
        let current = attendedHandle.getShowsAttended()
        for (index, status) in backup {
            attendedHandle.changeShowAttendedStatus(index: index, status: status, skipICloud: false)
        }
        for (index, _) in current where index.hasSuffix(":" + yearStr) && backup[index] == nil {
            attendedHandle.changeShowAttendedStatus(index: index, status: sawNoneStatus, skipICloud: false)
        }
        clearBackup(year: year)
    }
    
    /// Load backup for year, or nil if none.
    static func loadBackup(year: Int) -> [String: String]? {
        let url = backupFileURL(for: year)
        guard FileManager.default.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            return nil
        }
        return dict
    }
    
    static func clearBackup(year: Int) {
        try? FileManager.default.removeItem(at: backupFileURL(for: year))
    }
    
    /// Restore attendance to pre-AI state for the year. Uses backup; then clears backup and sets hasRunAI to false.
    /// - Returns: true if restore was performed, false if no backup.
    static func restore(attendedHandle: ShowsAttended, year: Int) -> Bool {
        guard let backup = loadBackup(year: year) else { return false }
        let yearStr = String(year)
        let current = attendedHandle.getShowsAttended()
        for (index, status) in backup {
            attendedHandle.changeShowAttendedStatus(index: index, status: status, skipICloud: false)
        }
        for (index, _) in current where index.hasSuffix(":" + yearStr) && backup[index] == nil {
            attendedHandle.changeShowAttendedStatus(index: index, status: sawNoneStatus, skipICloud: false)
        }
        clearBackup(year: year)
        setHasRunAI(for: year, value: false)
        return true
    }
}

File: andriod/app/src/main/java/com/Bands70k/AIScheduleStorage.java
package com.Bands70k;

import android.util.Log;

import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Persists "has run AI schedule" per year and backup of attendance for restore.
 * Port of Swift AIScheduleStorage.
 */
public class AIScheduleStorage {

    private static final String TAG = "AIScheduleStorage";
    private static final String HAS_RUN_PREFIX = "AIScheduleHasRun_";
    private static final String BACKUP_FILE_PREFIX = "AIScheduleBackup_";

    public static boolean hasRunAI(int year) {
        return SharedPreferencesManager.getInstance().getBoolean(HAS_RUN_PREFIX + year, false);
    }

    public static void setHasRunAI(int year, boolean value) {
        SharedPreferencesManager.getInstance().setBoolean(HAS_RUN_PREFIX + year, value);
    }

    private static File backupFileForYear(int year) {
        File dir = FileHandler70k.baseDirectory;
        if (!dir.exists()) dir.mkdirs();
        return new File(dir, BACKUP_FILE_PREFIX + year + ".json");
    }

    /**
     * Save attendance state for the given year (only keys ending with ":year"). Call before applying AI schedule.
     */
    public static void saveBackup(Map<String, String> attended, int year) {
        if (attended == null) return;
        String suffix = ":" + year;
        Map<String, String> filtered = new HashMap<>();
        for (Map.Entry<String, String> e : attended.entrySet()) {
            if (e.getKey() != null && e.getKey().endsWith(suffix)) {
                filtered.put(e.getKey(), e.getValue() != null ? e.getValue() : "");
            }
        }
        if (filtered.isEmpty()) return;
        try {
            JSONObject json = new JSONObject(filtered);
            File f = backupFileForYear(year);
            try (FileOutputStream out = new FileOutputStream(f)) {
                out.write(json.toString().getBytes(StandardCharsets.UTF_8));
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to save backup for year " + year + ": " + e.getMessage());
        }
    }

    /**
     * Saves year slice for wizard rollback (may be empty). Always writes a file so cancel-after-clear can restore.
     */
    public static void saveWizardRollbackBackup(Map<String, String> attended, int year) {
        if (attended == null) attended = new HashMap<>();
        String suffix = ":" + year;
        Map<String, String> filtered = new HashMap<>();
        for (Map.Entry<String, String> e : attended.entrySet()) {
            if (e.getKey() != null && e.getKey().endsWith(suffix)) {
                filtered.put(e.getKey(), e.getValue() != null ? e.getValue() : "");
            }
        }
        try {
            JSONObject json = new JSONObject(filtered);
            File f = backupFileForYear(year);
            try (FileOutputStream out = new FileOutputStream(f)) {
                out.write(json.toString().getBytes(StandardCharsets.UTF_8));
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to save wizard rollback backup for year " + year + ": " + e.getMessage());
        }
    }

    /**
     * Load backup for year, or null if none.
     */
    @SuppressWarnings("unchecked")
    public static Map<String, String> loadBackup(int year) {
        try {
            File f = backupFileForYear(year);
            if (!f.exists()) return null;
            try (FileInputStream in = new FileInputStream(f)) {
                byte[] buf = new byte[(int) f.length()];
                int n = in.read(buf);
                if (n <= 0) return null;
                JSONObject json = new JSONObject(new String(buf, 0, n, StandardCharsets.UTF_8));
                Map<String, String> out = new HashMap<>();
                Iterator<String> it = json.keys();
                while (it.hasNext()) {
                    String k = it.next();
                    out.put(k, json.optString(k, ""));
                }
                return out;
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to load backup for year " + year + ": " + e.getMessage());
            return null;
        }
    }

    public static void clearBackup(int year) {
        File f = backupFileForYear(year);
        if (f.exists()) f.delete();
    }

    /**
     * Restore attendance to pre-AI state for the year. Returns true if restore was performed.
     */
    public static boolean restore(showsAttended attendedHandle, int year) {
        Map<String, String> backup = loadBackup(year);
        if (backup == null || backup.isEmpty()) return false;
        attendedHandle.restoreFromBackup(year, backup);
        clearBackup(year);
        setHasRunAI(year, false);
        return true;
    }
}
